### PR TITLE
WALMA-99: Flaky UI tests

### PR DESCRIPTION
**Pull request by DemeSzabolcs**
[WALMA-99](https://lombiq.atlassian.net/browse/WALMA-99)

[WALMA-99]: https://lombiq.atlassian.net/browse/WALMA-99?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: Lombiq.Tests.UI/Services/SqlServerManager.cs [hunk 1]
```diff
@@ -41,9 +43,11 @@ public sealed class SqlServerManager : IAsyncDisposable
     private const string DbSnapshotName = "Database.bak";
 
     private static readonly PortLeaseManager _portLeaseManager;
+    private static readonly SemaphoreSlim _semaphore;
```
Comment (dministro) [inline Lombiq.Tests.UI/Services/SqlServerManager.cs]:
```suggestion
    private static readonly SemaphoreSlim _semaphore = new(1, 1);
```

File: Lombiq.Tests.UI/Services/SqlServerManager.cs [hunk 2]
```diff
@@ -59,6 +63,7 @@ static SqlServerManager()
     {
         var agentIndexTimesHundred = TestConfigurationManager.GetAgentIndexOrDefault() * 100;
         _portLeaseManager = new PortLeaseManager(13000 + agentIndexTimesHundred, 13099 + agentIndexTimesHundred);
+        _semaphore = new(1, 1);
```
Comment (dministro) [inline Lombiq.Tests.UI/Services/SqlServerManager.cs]:
```suggestion
```

File: Lombiq.Tests.UI/Services/SqlServerManager.cs [hunk 3]
```diff
@@ -114,131 +118,205 @@ public async Task<SqlServerRunningContext> CreateDatabaseAsync()
...
+                do
+                {
+                    // Copy back snapshot.
+                    await _docker.ExecuteAsync(
+                        CancellationToken.None,
+                        "cp",
+                        Path.Combine(local),
+                        $"{containerName}:{remote}");
+
+                    result = await DockerExecuteAndGetOutputAsync(
```
Comment (dministro) [inline Lombiq.Tests.UI/Services/SqlServerManager.cs]: In the case of `InvalidOperationException` thrown by `DockerExecuteAndGetOutputAsync` the retry logic does not apply.

Can we create a new method for this check? With a new method we can avoid the `S134` warning.
E.g.: `EnsureDockerFileExists`

File: Lombiq.Tests.UI/Services/SqlServerManager.cs [hunk 4]
```diff
@@ -114,131 +118,205 @@ public async Task<SqlServerRunningContext> CreateDatabaseAsync()
...
+                        containerName,
+                        "ls",
+                        remote[..remote.LastIndexOf('/')]);
```
Comment (dministro) [inline Lombiq.Tests.UI/Services/SqlServerManager.cs]:
```suggestion
                        Path.GetFileName(remote);
```

File: Lombiq.Tests.UI/Services/SqlServerManager.cs [hunk 5]
```diff
@@ -114,131 +118,205 @@ public async Task<SqlServerRunningContext> CreateDatabaseAsync()
...
+                }
+                while (result.IsNullOrEmpty() && retryCount < maxRetries + 1);
 
-        var restore = new Restore();
-        restore.Devices.AddDevice(GetSnapshotFilePath(snapshotDirectoryPathRemote), DeviceType.File);
-        restore.Database = _databaseName;
-        restore.ReplaceDatabase = true;
+                if (result.IsNullOrEmpty())
+                {
+                    throw new FileNotFoundException(
```
Comment (dministro) [inline Lombiq.Tests.UI/Services/SqlServerManager.cs]: Please create a custom exception for this. `FileNotFoundException` doesn't fit the real meaning of the context.
E.g.: `DockerFileCopyExcption` or similar.

**Review by dministro — CHANGES_REQUESTED** [review]

File: Lombiq.Tests.UI/Services/SqlServerManager.cs (hunk 4 above)
Comment (DemeSzabolcs) [reply Lombiq.Tests.UI/Services/SqlServerManager.cs]: I think this is wrong. It's on the docker so we can't use `Path.GetFileName(remote)`. UI test also fails with this.

Comment (dministro) [reply Lombiq.Tests.UI/Services/SqlServerManager.cs]: Yes, you are right. The directory is needed here and not the file name. Sorry.
Please use `Path.GetDirectoryName(remote)` instead.

**Review by dministro — APPROVED** [review]

**Pull request merged**